### ISSUE TITLE
MyImouto installation broken after moving to new server

### ISSUE THREAD

**Issue by tokumei** [myimouto/myimouto#160]
My home server was getting woefully out of date so a few days ago I installed a new operating system on it. One problem I've run into, though, is with MyImouto. I copied the installation from the old server to the new (the directory tree where the actual site is installed and the database), and it largely works, but seems to not be playing with MySQL entirely nicely.

Reading from the database seems to work in general. I can look at posts, search, look at comments and the wiki and the forum. The only thing I've noticed wrong in that respect is that if I try to look at the history log (i.e. by selecting Posts -> History) it gives me a 404 page (from MyImouto, not Apache). 

Writing is more problematic. The main issue is that I can't post new images, nor can I permanently delete images. When I try to upload a new image, I get the error "Couldn't retrieve new primary key." next to the Upload button. When I return to the main post index after this I get an error at the top of the screen reading "No service selected/no local service". If I attempt to delete an existing post, the first step (where the image is simply marked as deleted) works (including the "status" attribute in the posts table of the database for that row being updated to 'deleted'). If I attempt to permanently delete the image, it doesn't raise any error messages, but it also doesn't delete the image entry in the database. The actual image file does get deleted and every tag associated with the image has 1 subtracted from it (I can get tags down to negative numbers by doing this multiple times as well), but the image page still can be reached.

If I try to upload images with the import function, it just gives me the Kirino-with-torn-doll error image, deletes the images from the import directory, but doesn't add posts.

Editing a post's tags works, but raises an error at the top of the page simply reading "Error:". Despite the error the tags appear on the post page, but their counts are not updated (i.e. if I add an entirely new tag to an image its count will still be 0). Adding or removing posts from a pool does not work. Adding and deleting pools entirely does work however. 

Many other things that still involve writing to the database seem to work normally, however. Posting comments works, adding/editing notes works, creating posting to the forum works, editing wiki entries works. Adding new users and then logging in with them works. Adding and deleting artists works. Adding and deleting tag aliases and implications works. 

The new server is running PHP 5.6.9, MySQL 5.5.44, and Apache 2.4.10 (although I see no reason to suspect Apache has anything to do with this). I've also tried using MariaDB 10.0.20 but everything works the same. The old server was running PHP 5.5.8 for certain. I'm not 100% sure about the other parts but I believe it was running MySQL 5.5.35 and Apache 2.2.16.

None of these things I described leave errors in the log.

I know you don't necessarily have a lot of time to work on the project these days but any insight you might have into narrowing this problem down further would still be helpful. Thanks!

**Comment by gcscaglia** [comment]
These are just a few guesses, may as well be completely unrelated.

* You said you copied the database from the old install to the new one. If the versions you specified are correct, then the process is not a simple backup/restore; There is a [upgrade procedure](https://dev.mysql.com/doc/refman/5.6/en/upgrading.html) to be follow when crossing from 5.5 to 5.6. Never did one myself, so I can't be of much help in this case.

* I've had similar errors in \*write\* operations as well but the cause where the my.imouto folder permissions. There where mixed users and apache won't work correctly unless everything is owned by the same user the service is running as. When trying to write to some folders I would get permission denied on the logs.

* If all else fails, the my.imouto logs are on `/path/to/my.imouto/log`. Take a peek there, maybe the error messages will point you in the right direction.

**Comment by tokumei** [comment]
Thanks for the pointer to the application logs. I was just looking at the Apache logs for the MyImouto virtual host, and wasn't really aware of these different logs. Looking through them I found the actual errors being returned by the framework and determined that the actual problem is that the triggers for the database have the old database user as the definer, and I made a user with a different name on the new installation. I didn't know that could cause problems, whoops.

By changing the definer in those triggers to the new database user I seem to have solved the post/delete issues. I still get the "No service selected/no local service" error message, but now that I think about it I think I've seen that on the old installation as well. When I look at the source code it seems like that error is related to the similar image search functionality and has nothing in particular to do with this problem anyway.
